### ISSUE TITLE
ViewState: Type?

### ISSUE THREAD

**Issue by PaulWoitaschek** [sockeqwe/mosby#128]
Is there a reason that in `MvpViewStateFragment` `getViewState` does not expose its type? I don't like that I have to cast my view state implementations.

Why isn't the whole signature like:

``` java
public abstract class MvpViewStateFragment<V extends MvpView, P extends MvpPresenter<V>, VS extends ViewState<V>>
    extends MvpFragment<V, P> implements BaseMvpViewStateDelegateCallback<V, P> {
```

then all `ViewState<V>` could be replaced by `VS`.

**Comment by sockeqwe** [comment]
Didn't wanted to add more generic type definitions. I'm open to discuss that ...

I, personally, don't have a problem with casts and one could write his own base class with a generic type parameter for ViewState.

**Comment by PaulWoitaschek** [comment]
I think casts are to be avoided where possible. The power of the compiler helps you greatly to reduce bugs.
You could accidentally return the wrong type and get an class cast exception at runtime. 
I cleaned up a little and exposed at least the generic here:
https://github.com/sockeqwe/mosby/pull/129

Really adding the view state would require more work as there will be clashes with parent classes.

**Comment by PaulWoitaschek** [comment]
To explain what I mean:

I think it is very vital for a library that expects a specific implementation of a class to use generics correctly. Yes, you have to add generics. But it provides you with safety. 

There is a reason why you have to do that ugly `((MyPresenter) getPresenter())` cast. You are doing something potentially harmful.

Just returning a generic `ViewState` is like going back all the way down to Java 1.4 where you were just using `List` and always had to worry: Did I put `String` in here? Or `Integer`?

If this is wanted I can try to make a pull request and generify the `ViewState`. However please be aware that that affects many classes and **is** a breaking change as code written before won't compile any longer. But that might be even a good idea which can lead to bugs being detected.

**Comment by sockeqwe** [comment]
I'm willing to merge your pull request, I simply hadn't time yet (I'm sorry for not communicating this earlier). This will be part of `Mosby 3.0`. Since it's a major release, breaking API changes are ok. I have plans to continue the work on Mosby 3.0 at the end of April

**Comment by francoispluchino** [comment]
@sockeqwe Where are you with this issue, and more generally for the release of the stable version 3.0? Thanks!

**Comment by sockeqwe** [comment]
I don't have time to work on that before mid June ... I can't say when a stable version will be available. I hope mid July, but I can't make any promise.

**Comment by sockeqwe** [comment]
implemented in `3.0`
